File: bidding/bidding_nt.hs
import Bridge
import Data.Char (digitToInt)

type Hand = (HCP, HCP, Tricks)
type HCP = Int
type Tricks = Int
type Level = Int
type BiddingSequence = (HCP, HCP, [Level])
type Loss = Int

main :: IO ()
main = print . evalSystem . getHands =<< readFile "Library.asc"

getHands :: String -> [Hand]
getHands = map getHand . lines

getHand :: String -> Hand
getHand xs = (hcphigher, hcplower, tricks)
  where
    calchcp = sum . map atohcp
    cards = take 68 xs
    hcp1 = calchcp $ words cards !! 1
    hcp2 = calchcp $ words cards !! 3
    hcphigher = if hcp1 >= hcp2 then hcp1 else hcp2
    hcplower = if hcp1 >= hcp2 then hcp2 else hcp1
    tricks = digitToInt (xs !! 69)

evalSystem :: [Hand] -> Loss
evalSystem = sum . map eval

eval :: Hand -> Loss
eval (hcp1, hcp2, tricks) = loss (bidding hcp1 hcp2) tricks

loss :: Level -> Tricks -> Loss
loss level tricks = downs + loses
  where
    downs = if level > 0 then max (level + 6 - tricks) 0 else 0
    loses  = gamelose + partiallose
    gamelose = if (level < 3 && tricks >= 9) then 1 else 0
    partiallose = if (level < 1 && tricks >= 7) then 1 else 0

biddingSystem :: [Level] -> HCP
biddingSystem [3] = 24
biddingSystem (3:xs) = 25 - biddingSystem xs
biddingSystem [2] = 21
biddingSystem (2:xs) = 24 - biddingSystem xs
biddingSystem [1] = 15
biddingSystem _ = undefined

bidding :: HCP -> HCP -> Level
bidding hcp1 hcp2 = takeLevel biddingSequence
  where
    biddingSequence = until isPass bid (hcp1, hcp2, [])
    takeLevel (_, _, xs) = if length xs >= 2 then xs !! 1 else 0
    isPass (_, _, xs) = xs /= [] && head xs == 0

bid :: BiddingSequence -> BiddingSequence
bid (hcp1, hcp2, xs) = if nextxs /= [] then (hcp2, hcp1, head nextxs) else (hcp2, hcp1, 0:xs)
  where
    currentLevel = if xs == [] then 0 else head xs
    nextxs = [i:xs | i <- [3,2..currentLevel+1], hcp1 >= biddingSystem (i:xs)]

File: bidding/bidding_nt.c
#include <stdio.h>
#include <stdlib.h>

int loss_deal(int, int, int);
int bidding(int, int, int);
void gethand(int[], int[], int[]);
int atohcp(char a);

const int deals = 717102;
int hcp1[deals], hcp2[deals], tricks[deals];

int main()
{
    int loss_total = 0;
    gethand(hcp1, hcp2, tricks);
    for (int i = 0; i < deals; i++) {
        loss_total += loss_deal(hcp1[i], hcp2[i], tricks[i]);
    }
    printf("%d\n", loss_total);
    return 0;
}

// そのディールでのロス
int loss_deal(int hcp1, int hcp2, int tricks)
{
    int loss = 0, level = bidding(hcp1, hcp2, 0);
    if (tricks < level + 6 && level > 0)
        loss += level + 6 - tricks;
    if (tricks >= 7 && level < 1)
        loss++;
    if (tricks >= 9 && level < 3)
        loss++;
    return loss;
}

// ビッド（返り値はレベル）
int bidding(int hcp1, int hcp2, int sequence)
{
    const int system[8] = { 0, 15, 21, 9, 24, 10, 4, 16 };
    if (hcp1 >= system[sequence + 4])
        return 3;
    if (sequence >= 2)
        return 2;
    if (hcp1 >= system[sequence + 2])
        return bidding(hcp2, hcp1, sequence + 2);
    if (sequence >= 1)
        return 1;
    if (hcp1 >= system[sequence + 1])
        return bidding(hcp2, hcp1, sequence + 1);
    return 0;
}

// Library.ascからHCPとトリックを取得
void gethand(int hcp1[], int hcp2[], int tricks[])
{
    FILE *data = fopen("Library.asc", "r");
    for (int i = 0; i < deals; i++) {
        for (int j = 0; j < 17; j++)
            getc(data);
        for (int j = 0; j < 17; j++)
            hcp1[i] += atohcp(getc(data));
        for (int j = 0; j < 17; j++)
            getc(data);
        for (int j = 0; j < 17; j++)
            hcp2[i] += atohcp(getc(data));
        if (fscanf(data, "%*1c%1x%*19c", &tricks[i]) != 1)
            abort();
        if (hcp1[i] < hcp2[i]) {
            int w = hcp1[i];
            hcp1[i] = hcp2[i];
            hcp2[i] = w;
        }
    }
    fclose(data);
}

// 文字をHCPに変換
int atohcp(char a)
{
    switch (a) {
    case 'A':
        return 4;
    case 'K':
        return 3;
    case 'Q':
        return 2;
    case 'J':
        return 1;
    default:
        return 0;
    }
}
